broadinstitute/single_cell_portal_core
# Reuse expression data for dotplots and heatmaps (SCP-5263)

#### BACKGROUND
Currently, when a user is visualizing multi-gene expression data as a dotplot or heatmap, we make separate API calls from each component to load the same data.  This means that one gene search spawn 4 calls for identical data (2 for expression as a tsv, 2 for annotations as a tsv).  Since the data is identical, we should be able to share it across components.  Futhermore, the current setup can lead to stability issues when users are requesting many genes in rapid succession, as each new gene requested spawns 4 more API calls.

#### CHANGES
This update changes both the UI and API layers to reuse expression data for both components.  This is done by switching from a GCT format (plain text matrix file) to a JSON representation of a [Morpheus dataset](https://software.broadinstitute.org/morpheus/configuration.html#datasetJSON) via the API which is then passed from the `DotPlot` component to the associated `HeatMap` component.  Additionally, this dataset contains both expression and annotation information, meaning that we only make 1 API call instead of 4.  This will reduce load on the server, and could open avenues down the road for increasing the limit of the number of genes requested in a single search, or client-side caching of individual gene expression scores.  Benchmarking has shown that this is at least as performant as the current implementation, and potentially 10-20% faster.  Additionally, this new implementation works with the refinements made in #1842 so that requests can be saved in the service worker cache for faster iteration times in local development.

#### MANUAL TESTING
1. Boot as normal and load the `Human milk - differential expression` study
2. Select the `All Cells UMAP` cluster and `General_Celltype` annotation
3. Open Chrome DevTools -> Network
4. Query for the genes `STAT5A` and `CD68`
5. In the network tab, confirm you see only one request for the expression JSON endpoint like: `https://localhost:3000/single_cell/api/v1/studies/SCP66/expression/json?annotation_name=General_Celltype&annotation_scope=study&annotation_type=group&cluster=All%20Cells%20UMAP&genes=STAT5A%2CCD68&subsample=all`
6. Open the dotplot tab and confirm you see the following plot:
![2_genes dotplot](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/de8f5c51-282d-41f6-b7e9-76f9a13ea14c)
7. Click the heatmap tab and confirm there is a heatmap for both genes
8. Go back to the dotplot tab, and request another gene: `KLF6`
9. Confirm that the dotplot starts to re-render, and only one more request is made to the expression JSON endpoint in the network tab
10. Confirm the following dotplot finishes rendering with the plot below, and that the heatmap tab now has 3 genes
![3_genes dotplot](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/5b060ea3-85ce-4ad0-8fe8-6e42a5f8427c)
